### DISCUSSION TITLE
Sending Additional Data from Arduino/TC4 to Artisan

### DISCUSSION THREAD

**Discussion by bfailing** [artisan-roaster-scope/artisan#970] [Q&A] [answered]
I want to send additional data (e.g., variable values that are either doubles or integers) from the TC4 to Artisan at the designated sampling interval (e.g., once per second along with temp values for the 4 channels).  And, ideally, I'd like this additional data to be exported by Artisan along with the other data such as temp values, heater level, fan level, etc.

Can Artisan be configured to allow this?  If so, how?

**Comment by MAKOMO** [comment, marked as answer]
No. You plan to change/extend the TC4 protocol. For this you need to implement that change on both sides, the TC4 firmware as well as Artisan. The protocol itself is not configurable on the Artisan side.

**Reply by MAKOMO** [reply]
While you are modifying the TC4 firmware, you could have the TC4 run a MODBUS server and configure Artisan to read the data from the corresponding registers. There is a MODBUS implementation for Arduino. Artisan can read from up to 8 MODBUS registers and write to any number of registers.

**Reply by bfailing** [reply]
What about using an extra device?  Is there a way, e.g., to emulate an extra device and send data via the emulated extra device?

**Reply by MAKOMO** [reply]
Well, Artisan reads up to 8 readings from the TC4 via the TC4/PID protocol.

chan1-4: the for temperatures
chan5/6: heater/fan duty
chan7/8: PID SV / TC4-internal temp

Maybe you can abuse one of those?

**Reply by bfailing** [reply]
I currently only have the fourth temp channel free, but already use it to send other data.  An extra channel or two would be helpful.

**Reply by MAKOMO** [reply]
Just "abuse" the chan 5/6 and/or the chan 7/8 or add two channels on both sides, Artisan and TC4.

**Reply by bfailing** [reply]
How do you add channels in Artisan?  Does it require modifying the Artisan code?

**Reply by MAKOMO** [reply]
Yes

**Reply by bfailing** [reply]
Is the process for building the Windows version of Artisan documented somewhere?

**Reply by MAKOMO** [reply]
All the build scripts are in the repository, however, there is no specific documentation for the build process. Awaiting contributions;)
